adapters/src/main/java/com/squareup/moshi/adapters/RuntimeJsonAdapterFactory.java

```diff
@@ -63,112 +75,117 @@
       throw new IllegalArgumentException(
           "The base type must not be Object. Consider using a marker interface.");
     }
-    return new RuntimeJsonAdapterFactory<>(baseType, labelKey);
-  }
-
-  RuntimeJsonAdapterFactory(Class<T> baseType, String labelKey) {
-    this.baseType = baseType;
-    this.labelKey = labelKey;
+    return new RuntimeJsonAdapterFactory<>(
+        baseType, labelKey, Collections.<String>emptyList(), Collections.<Type>emptyList());
   }
 
   /**
-   * Register the subtype that can be created based on the label. When an unknown type is found
+   * Returns a new factory that decodes instances of {@code subtype}. When an unknown type is found
    * during encoding an {@linkplain IllegalArgumentException} will be thrown. When an unknown label
    * is found during decoding a {@linkplain JsonDataException} will be thrown.
    */
-  public RuntimeJsonAdapterFactory<T> registerSubtype(Class<? extends T> subtype, String label) {
+  public RuntimeJsonAdapterFactory<T> withSubtype(Class<? extends T> subtype, String label) {
     if (subtype == null) throw new NullPointerException("subtype == null");
     if (label == null) throw new NullPointerException("label == null");
-    if (labelToType.containsKey(label) || labelToType.containsValue(subtype)) {
+    if (labels.contains(label) || subtypes.contains(subtype)) {
       throw new IllegalArgumentException("Subtypes and labels must be unique.");
     }
-    labelToType.put(label, subtype);
-    return this;
+    List<String> newLabels = new ArrayList<>(labels);
+    newLabels.add(label);
+    List<Type> newSubtypes = new ArrayList<>(subtypes);
+    newSubtypes.add(subtype);
+    return new RuntimeJsonAdapterFactory<>(baseType, labelKey, newLabels, newSubtypes);
   }
 
   @Override
   public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
     if (Types.getRawType(type) != baseType || !annotations.isEmpty()) {
       return null;
     }
-    int size = labelToType.size();
-    Map<String, JsonAdapter<Object>> labelToAdapter = new LinkedHashMap<>(size);
-    Map<Type, String> typeToLabel = new LinkedHashMap<>(size);
-    for (Map.Entry<String, Type> entry : labelToType.entrySet()) {
-      String label = entry.getKey();
-      Type typeValue = entry.getValue();
-      typeToLabel.put(typeValue, label);
-      labelToAdapter.put(label, moshi.adapter(typeValue));
+
+    List<JsonAdapter<Object>> jsonAdapters = new ArrayList<>();
```

Comment:
tiny nit: could be `new ArrayList<>(size)`

Reply:
Can fix in follow-up!